packages/protocol/contracts/ACE/noteRegistry/proxies/AdminUpgradeabilityProxy.sol

```diff
@@ -0,0 +1,24 @@
+pragma solidity ^0.5.0;
+
+import './BaseAdminUpgradeabilityProxy.sol';
+
+/**
+ * @title AdminUpgradeabilityProxy
+ * @dev Extends from BaseAdminUpgradeabilityProxy with a constructor for
+ * initializing the implementation, admin, and init data.
+ */
+contract AdminUpgradeabilityProxy is BaseAdminUpgradeabilityProxy, UpgradeabilityProxy {
+  /**
+   * Contract constructor.
+   * @param _logic address of the initial implementation.
+   * @param _admin Address of the proxy administrator.
+   * @param _data Data to send as msg.data to the implementation to initialize the proxied contract.
+   * It should include the signature and the parameters of the function to be called, as described in
+   * https://solidity.readthedocs.io/en/v0.4.24/abi-spec.html#function-selector-and-argument-encoding.
+   * This parameter is optional, if no data is given the initialization call to proxied contract will be skipped.
+   */
+  constructor(address _logic, address _admin, bytes memory _data) UpgradeabilityProxy(_logic, _data) public payable {
+    assert(ADMIN_SLOT == bytes32(uint256(keccak256('eip1967.proxy.admin')) - 1));
```

Comment:
Out of interest, when we are generating a location for the storage slots by hashing a string - why do we subtract 1 off the result? 

Presume it's just to increase the randomness of the resulting storage slot and make a collision even less likely?

Reply:
Detailed in the EIP 1967 spec (https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1967.md), the -1 is to reduce the chance of a pre-image attack

Reply:
Makes sense